walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,124 @@
+		MeetingRequest request) {
+		int duration = (int) request.getDuration();
+		Collection<String> attendees = request.getAttendees();
+		Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+		Collection<TimeRange> availableTimesWithOptionals = new ArrayList<>();
+
+		availableTimes = getAvaliableTimes(attendees, events, duration);
+
+		if (!optionalAttendees.isEmpty()) {
+			availableTimesWithOptionals = getAvaliableTimes(optionalAttendees,
+				events, duration);
+		}
+
+		if (optionalAttendees.isEmpty()) {
+			return availableTimes;
+		}
+
+		if (attendees.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		availableTimesWithOptionals = intersection(availableTimesWithOptionals,
+			availableTimes);
+
+		if (!availableTimesWithOptionals.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		return availableTimes;
+	}
+
+	/**
+	 * Finds avaliable times for the given atteendee list, input 
+	 * collection and event duration
+	 */
+	private Collection<TimeRange> getAvaliableTimes(Collection<String> attendees,
+		Collection<Event> events, int duration) {
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+
+		int start = 0;
+		int end = start + duration;
+		int availablePeriodStart = -1, availablePeriodEnd = -1;
+
+		while (end<= END_OF_DAY) {
+			TimeRange currentTimeRange = TimeRange.fromStartDuration(start, duration);
+			boolean works = true;
+
+			for (Event event: events) {
+				if (!works) break;
+
+				TimeRange eventTimeRange = event.getWhen();
+				if (currentTimeRange.overlaps(eventTimeRange)) {
+					Collection<String> eventAttendees = event.getAttendees();
+					if (!Collections.disjoint(attendees, eventAttendees)) {
+						works = false;
+						continue;
+					}
+				}
+			}
+
+			if (works) {
+				if (availablePeriodEnd == -1) {
+					availablePeriodStart = start;
+				}
+
+				availablePeriodEnd = end;
+
+				if (end == END_OF_DAY) {
+					TimeRange availableTime = TimeRange
+						.fromStartEnd(availablePeriodStart, availablePeriodEnd, false);
+					availableTimes.add(availableTime);
+				}
+			} else {
+				if (availablePeriodEnd != -1) {
+					TimeRange availableTime = TimeRange
+						.fromStartEnd(availablePeriodStart, availablePeriodEnd, false);
+					availableTimes.add(availableTime);
+				}
+
+				availablePeriodStart = -1;
+				availablePeriodEnd = -1;
+			}
+			start++;
+			end++;
+		}
+
+		return availableTimes;
+	}
+
+	/**
+	 * Finds intersection between elements of two lists of 
+	 * TimeRange and returns as a list
+	 */
+	private List<TimeRange> intersection(Collection<TimeRange> list1,
+		Collection<TimeRange> list2) {
+		List<TimeRange> intersection = new ArrayList<>();
+
+		for (TimeRange el1: list1) {
+			for (TimeRange el2: list2) {
+				if (el1.contains(el2)) {
+					intersection.add(el2);
+				} else if (el2.contains(el1)) {
+					intersection.add(el1);
+				}
+			}
+		}
+
+		return intersection;
```

Comment:
There is an O(n) way using sets to get the intersection of two collections, vs O(n^2) here.

https://stackoverflow.com/questions/38768997/java-collection-methods-for-union-intersection-and-set-difference

Reply:
Here I am not only looking for intersection between two lists but intersection between TimeRange objects times. For instance, if one list has element TimeRange(8:00 - 10:00) and the other one has TimeRange(9:00-10:00), since second TimeRange is contained into first one, I want to  include  TimeRange(9:00-10:00) in intersection, even though two objects are not equivalent. That's why I think I need to check every pair.

However, I still probably going to end up changing the method I as I optimize the algorithm.

Reply:
I see, thank you for explaining! That makes senes.